File: CI1056 - Algoritmos 2/Binário/BuscaBináriaRec.c
#include <stdio.h>

#define TAM 200

// Função para ler os elementos do vetor
void lerVetor(int n, int v[]) 
{
    for (int i = 0; i < n; i++) 
    {
        scanf("%d", &v[i]);
    }
}

// Função recursiva para realizar a busca binária
int PesqBinRecursiva(int v[], int inicio, int fim, int elemento) 
{
    if (inicio > fim) 
    {
        return 0; // Elemento não encontrado
    }

    int meio = (inicio + fim) / 2;

    if (elemento == v[meio]) 
    {
        return meio; // Retorna o índice do elemento encontrado
    }
    else if (elemento > v[meio]) 
    {
        return PesqBinRecursiva(v, meio + 1, fim, elemento); // Busca na metade direita
    } 
    else 
    {
        return PesqBinRecursiva(v, inicio, meio - 1, elemento); // Busca na metade esquerda
    }
}

int main() 
{
    int v[TAM];
    int n, elemento, resultado;

    scanf("%d", &n);
    lerVetor(n, v);
    scanf("%d", &elemento);

    // Realiza a busca binária recursiva e imprime o resultado
    resultado = PesqBinRecursiva(v, 0, n - 1, elemento);
    
    if (resultado != 0) 
    {
        printf("%d\n", resultado);
    } 
    else 
    {
        printf("0\n");
    }

    printf("\n");
    return 0;
}

File: CI1056 - Algoritmos 2/Binário/BuscaBináriaRec.pas
program recursao;

type
    vetor = array [1..200] of integer;

var
    v: vetor;
    n, elemento: integer;

procedure leVetor (n: integer; var v: vetor);
var
    i: integer;
begin
    for i := 1 to n do
        read(v[i]);
end;

function PesqBinRec (elemento: integer; var v: vetor; inicio, fim: integer): integer;
var
    meio: integer;
begin
    if (inicio > fim) then
        PesqBinRec := 0
    else
    begin
        meio := (inicio + fim) div 2;
        if (elemento = v[meio]) then
            PesqBinRec := meio
        else if (elemento < v[meio]) then
            PesqBinRec := PesqBinRec(elemento, v, inicio, meio - 1)
        else
            PesqBinRec := PesqBinRec(elemento, v, meio + 1, fim);
    end;
end;

begin
    read(n, elemento);
    leVetor(n, v);
    if (PesqBinRec(elemento, v, 1, n) <> 0) then
        writeln(PesqBinRec(elemento, v, 1, n))
    else
        writeln('Elemento não encontrado');
end.